### ISSUE TITLE
Eigen headers still #included with OpenBLAS

### ISSUE THREAD

**Issue by christoph-conrads** [facebookarchive/caffe2#1355]
I compiled commit 55b3b14101fc75fc3564ef2887b1b689039298a3 from source with `-DBLAS=OpenBLAS` and without CUDA. Yet, some headers still `#include` Eigen headers causing a compilation error when you want to use the C++ API. Since I use OpenBLAS, I expect that there is *no* Eigen dependency when using the Caffe2 headers.

Minimum working example:
```c++
#include <caffe2/core/predictor.h>
int main() {}
```
Error message:
```
christoph:/tmp$ g++ mwe.cpp -I/tmp/caffe2.x86_64-linux-gnu/include -c -std=c++11
In file included from /tmp/caffe2.x86_64-linux-gnu/include/caffe2/core/context.h:30:0,
                 from /tmp/caffe2.x86_64-linux-gnu/include/caffe2/core/tensor.h:30,
                 from /tmp/caffe2.x86_64-linux-gnu/include/caffe2/core/net.h:34,
                 from /tmp/caffe2.x86_64-linux-gnu/include/caffe2/core/predictor.h:19,
                 from mwe.cpp:1:
/tmp/caffe2.x86_64-linux-gnu/include/caffe2/utils/math.h:36:22: fatal error: Eigen/Core: No such file or directory
compilation terminated.
```
[Line 36 in `caffe2/utils/math.h`](https://github.com/caffe2/caffe2/blob/55b3b14101fc75fc3564ef2887b1b689039298a3/caffe2/utils/math.h#L36)

**Comment by pietern** [comment]
Thanks for the report.

**Comment by Yangqing** [comment]
Thanks for the report - we'll do some passes to clear up the header inclusions, especially the public ones that a dependent library may need to include.

**Comment by zeakey** [comment]
I run into another error with `-DBLAS=OpenBLAS`.
```
In file included from /usr/include/Eigen/src/Core/ArrayBase.h:111:0,
                 from /usr/include/Eigen/Core:350,
                 from /packages/caffe2/caffe2/utils/math.h:36,
                 from /packages/caffe2/caffe2/core/context.h:30,
                 from /packages/caffe2/caffe2/operators/cross_entropy_op.h:20,
                 from /packages/caffe2/caffe2/operators/cross_entropy_op.cc:17:
/usr/include/eigen3/Eigen/src/plugins/MatrixCwiseBinaryOps.h: In instantiation of ‘const Eigen::CwiseBinaryOp<Eigen::internal::scalar_max_op<typename Eigen::internal::traits<T>::Scalar>, const Derived, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<typename Eigen::internal::traits<T>::Scalar>, Derived> > Eigen::ArrayBase<Derived>::cwiseMax(const Scalar&) const [with Derived = Eigen::Map<const Eigen::Array<float, -1, -1>, 0, Eigen::Stride<0, 0> >; typename Eigen::internal::traits<T>::Scalar = float; Eigen::ArrayBase<Derived>::Scalar = float]’:
/packages/caffe2/caffe2/operators/cross_entropy_op.cc:300:73:   required from here
/usr/include/eigen3/Eigen/src/plugins/MatrixCwiseBinaryOps.h:110:72: error: could not convert ‘Eigen::ArrayBase<Derived>::cwiseMax(const Eigen::ArrayBase<OtherDerived>&) const [with OtherDerived = Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<float>, Eigen::Array<float, -1, -1> >; Derived = Eigen::Map<const Eigen::Array<float, -1, -1>, 0, Eigen::Stride<0, 0> >; typename Eigen::internal::traits<T>::Scalar = float]((*(const Eigen::ArrayBase<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<float>, Eigen::Array<float, -1, -1> > >*)(& Eigen::DenseBase<Derived>::Constant(Eigen::DenseBase<Derived>::Index, Eigen::DenseBase<Derived>::Index, const Scalar&) [with Derived = Eigen::Array<float, -1, -1>; Eigen::DenseBase<Derived>::ConstantReturnType = Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<float>, Eigen::Array<float, -1, -1> >; typename Eigen::internal::traits<T>::Scalar = float; Eigen::DenseBase<Derived>::Index = long int; Eigen::DenseBase<Derived>::Scalar = float](((const Eigen::EigenBase<Eigen::Map<const Eigen::Array<float, -1, -1>, 0, Eigen::Stride<0, 0> > >*)this)->Eigen::EigenBase<Derived>::cols<Eigen::Map<const Eigen::Array<float, -1, -1>, 0, Eigen::Stride<0, 0> > >(), (* & other)))))’ from ‘const Eigen::CwiseBinaryOp<Eigen::internal::scalar_max_op<float>, const Eigen::Map<const Eigen::Array<float, -1, -1>, 0, Eigen::Stride<0, 0> >, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<float>, Eigen::Array<float, -1, -1> > >’ to ‘const Eigen::CwiseBinaryOp<Eigen::internal::scalar_max_op<float>, const Eigen::Map<const Eigen::Array<float, -1, -1>, 0, Eigen::Stride<0, 0> >, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<float>, Eigen::Map<const Eigen::Array<float, -1, -1>, 0, Eigen::Stride<0, 0> > > >’
   return cwiseMax(Derived::PlainObject::Constant(rows(), cols(), other));
```

My environment is: `Ubuntu 14` with `Eigen3`. I'm Compiling with <https://github.com/caffe2/caffe2/commit/003c7e0e069665a078803a5a4efb4c33d36208b6>

**Comment by pietern** [comment]
The fix was merged in 018e952a160542d35ebfc6b2377f5088ed197cb5.

@zeakey Can you open a new issue for your problem?
